src/smt.cpp

```diff
@@ -1160,6 +1160,73 @@ Expr &Expr::operator|=(const Expr &rhs) {
   return *this;
 }
 
+EXPR_BVOP_UINT64(shl)
+Expr Expr::shl(const Expr &rhs) const {
+  CHECK_LOCK2(rhs);
+
+  uint64_t a, b;
+  if (isUInt(a) && rhs.isUInt(b))
+    return mkBV(a << b, rhs.bitwidth());
+  else if (rhs.isUInt(b)) {
+    if (b == 0)
+      return *this;
+  } else if (isUInt(a)) {
+    if (a == 0)
+      return *this;
+  }
+
+  Expr e;
+  SET_Z3_USEOP(e, rhs, shl);
+  SET_CVC5_USEOP(e, rhs, BITVECTOR_SHL);
+  return e;
+}
+
+EXPR_BVOP_UINT64(ashr)
+Expr Expr::ashr(const Expr &rhs) const {
+  CHECK_LOCK2(rhs);
+
+  uint64_t a, b;
+  // The value of a >> b for unsigned a is the integer part of a/(2^b)
+  // which is equivalent to lshr.
+  // Therefore we cannot apply uint optimization for ashr.
+  // Even if we use signed int, the value of >> on a signed int
+  // is implementation-defined until c++17 (it is ashr since c++20),
+  // so it is hard to implement a robust optimization.
+  if (rhs.isUInt(b)) {
+    if (b == 0)
+      return *this;
+  } else if (isUInt(a)) {
+    if (a == 0)
+      return *this;
+  }
+
+  Expr e;
+  SET_Z3_USEOP(e, rhs, ashr);
+  SET_CVC5_USEOP(e, rhs, BITVECTOR_ASHR);
+  return e;
+}
+
+EXPR_BVOP_UINT64(lshr)
+Expr Expr::lshr(const Expr &rhs) const {
+  CHECK_LOCK2(rhs);
+
+  uint64_t a, b;
+  if (isUInt(a) && rhs.isUInt(b))
+    return mkBV(a >> b, rhs.bitwidth());
```

Comment:
Clang lowers `a >> b` into `lshr` and `lshr` produces poison if `b` is equal to or larger than bitwidth (https://llvm.org/docs/LangRef.html#lshr-instruction). I will add `&& b < 64` at the if condition.
Also same for shl.